### PR TITLE
Supporting sort by multiple columns

### PR DESCRIPTION
**Pull request by chathurawidanage**

### CODE REVIEW

**Comment by nirandaperera** [comment]
@chathurawidanage I think the tests are failing because the csv1_0.csv has been changed! :-) could you please change that?

**Comment by chathurawidanage** [comment]
Yeah. I tried to revert that based on what's available in commit history.. But not sure whether there was two 41s in the original file.

**Comment by vibhatha** [comment]
Can we add some Python test cases for new options?

File: cpp/src/cylon/arrow/arrow_comparator.cpp [hunk 1]
```diff
@@ -16,14 +16,14 @@
...
+template <typename ARROW_TYPE>
 class NumericArrowComparator : public ArrowComparator {
   int compare(const std::shared_ptr<arrow::Array> &array1, int64_t index1,
               const std::shared_ptr<arrow::Array> &array2, int64_t index2) override {
     auto reader1 = std::static_pointer_cast<arrow::NumericArray<ARROW_TYPE>>(array1);
     auto reader2 = std::static_pointer_cast<arrow::NumericArray<ARROW_TYPE>>(array2);
-
-    return (int) (reader1->Value(index1) - reader2->Value(index2));
+    auto diff = reader1->Value(index1) - reader2->Value(index2);
+    return (diff > 0) - (diff < 0);
```
Comment (nirandaperera) [inline cpp/src/cylon/arrow/arrow_comparator.cpp:26]: shall we add a comment here with a link to the bit twiddling hacks site?
https://graphics.stanford.edu/~seander/bithacks.html#CopyIntegerSign

File: cpp/src/cylon/arrow/arrow_comparator.cpp [hunk 2]
```diff
@@ -111,71 +147,111 @@ int TableRowComparator::compare(const std::shared_ptr<arrow::Table> &table1, int
...
+    return CreateArrayIndexComparatorUtil<false>(array);
   }
 }
 
 RowComparator::RowComparator(std::shared_ptr<cylon::CylonContext> &ctx,
-                             const std::shared_ptr<arrow::Table> *tables,
-                             int64_t *eq,
+                             const std::shared_ptr<arrow::Table> *tables, int64_t *eq,
                              int64_t *hs) {
   this->tables = tables;
```
Comment (nirandaperera) [inline cpp/src/cylon/arrow/arrow_comparator.cpp:256]: shall we move these to the initializer list?

File: cpp/src/cylon/arrow/arrow_kernels.cpp [hunk 3]
```diff
@@ -181,80 +178,91 @@ using DoubleArraySplitter = ArrowArrayNumericSplitKernel<arrow::DoubleType>;
...
   using T = typename TYPE::c_type;
 
-  NumericIndexSortKernel(arrow::MemoryPool *pool, bool ascending) : IndexSortKernel(pool, ascending) {}
+  NumericIndexSortKernel(arrow::MemoryPool *pool, bool ascending)
+      : IndexSortKernel(pool, ascending) {}
 
   arrow::Status Sort(const std::shared_ptr<arrow::Array> &values,
                      std::shared_ptr<arrow::UInt64Array> &offsets) const override {
     auto array = std::static_pointer_cast<arrow::NumericArray<TYPE>>(values);
+    array->GetView(0);
```
Comment (nirandaperera) [inline cpp/src/cylon/arrow/arrow_kernels.cpp:255]: I believe this is not required

File: cpp/src/cylon/arrow/arrow_kernels.hpp [hunk 4]
```diff
@@ -129,6 +131,12 @@ arrow::Status SortIndicesInPlace(arrow::MemoryPool *memory_pool,
                                  std::shared_ptr<arrow::Array> &values,
                                  std::shared_ptr<arrow::UInt64Array> &offsets);
 
+arrow::Status SortIndicesMultiColumns(arrow::MemoryPool *memory_pool,
+                                      const std::shared_ptr<arrow::Table> &table,
+                                      const std::vector<int64_t> &columns,
+                                      std::shared_ptr<arrow::UInt64Array> &offsets,
+                                      const std::vector<bool> &ascending);
```
Comment (nirandaperera) [inline cpp/src/cylon/arrow/arrow_kernels.hpp:138]: should we add a default value for ascending here?

**Review by nirandaperera — CHANGES_REQUESTED** [review]

**Comment by nirandaperera** [comment]
@chathurawidanage from hereon, let's not send PRs from your master branch. It could get really ugly because we are doing  squash and merge :slightly_smiling_face:

**Review by nirandaperera — APPROVED** [review]

**Pull request merged**